cpp/src/structs/utilities.cpp

```
@@ -168,15 +167,132 @@ flatten_nested_columns(table_view const& input,
+                                               cudf::column_view const& blueprint)
+{
+  // "Consume" columns from `flattened`, starting at `current_index`,
+  // based on the provided `blueprint` struct col. Recurse for struct children.
+  CUDF_EXPECTS(blueprint.type().id() == type_id::STRUCT,
+               "Expected blueprint column to be a STRUCT column.");
+
+  CUDF_EXPECTS(current_index < flattened.size(), "STRUCT column can't have 0 children.");
+
+  auto const num_rows = flattened[current_index]->size();
+
+  // cudf::flatten_nested_columns() executes depth first, and serializes the struct null vector
+  // before the child/member columns.
+  // E.g. STRUCT_1< STRUCT_2< A, B >, C > is flattened to:
+  //      1. Null Vector for STRUCT_1
+  //      2. Null Vector for STRUCT_2
+  //      3. Member STRUCT_2::A
+  //      4. Member STRUCT_2::B
+  //      5. Member STRUCT_1::C
+  //
+  // Extract null-vector *before* child columns are constructed.
+  auto struct_null_column_contents = flattened[current_index++]->release();
+
+  auto struct_members = vector_of_columns{};
+  struct_members.reserve(blueprint.num_children());
+
+  std::transform(blueprint.child_begin(),
+                 blueprint.child_end(),
+                 std::back_inserter(struct_members),
+                 unflattener{flattened, current_index});
+
+  return cudf::make_structs_column(num_rows,
+                                   std::move(struct_members),
+                                   UNKNOWN_NULL_COUNT,  // Do count?
+                                   std::move(*struct_null_column_contents.null_mask));
+}
+
+bool is_struct(cudf::column_view const& col) { return col.type().id() == type_id::STRUCT; }
+
+/**
+ * @brief Check whether the specified column is of type LIST,
+ *        or any LISTs in its descendent columns.
+ *
+ */
+bool is_or_has_lists(cudf::column_view const& col)
+{
+  auto is_list = [](cudf::column_view const& col) { return col.type().id() == type_id::LIST; };
```

If this lambda is used only once, then we may not need it.

---

Yep, it's only used once. But it makes the calling code more readable.

---

I've now changed the `flatten_nested_columns()` code to check the same condition. It now fails on `LIST` columns, instead of failing silently.